### ISSUE TITLE
An unexpected mailsync error occurred (3221225477)

### ISSUE THREAD

**Issue by Zangoran** [Foundry376/Mailspring#292]
##### Are there any related issues?

I have searched using the supplied error code and I was unable to find any information on it.

##### What operating system are you using?
Microsoft Windows [Version 10.0.14393]
##### What version of Mailspring are you using?
![image](https://user-images.githubusercontent.com/7672085/32303947-d008e8ac-bf39-11e7-9a3e-8a788183c63f.png)

--

**Bug?**
##### Do you have any third-party plugins installed? If so, which ones?
None

##### Is the issue related to a specific email provider (Gmail, Exchange, etc.)?
Gmail

##### Is the issue reproducible with a particular attachment, message, signature, etc?

This only is occurring with my school issued email account. I believe that they have a script that removes all apps that are connected to my google account, which is breaking the link between my account and mailspring. The error comes in when I try to Re-authenticate by using the built in button in the accounts menu in mailspring. When I do this, I get the error code: 3221225477. Any help would be greatly appreciated.

**Comment by zopsi** [comment]
Same Error with GSuite. It seems to think all my other accounts are having issues after I get this too.

`An unexpected mailsync error occurred (3221225477)`

This keeps showing up in the logs

`[background] [error] syncFolderUIDRange - Fetching full headers for 1`

Operating System: Microsoft Windows 10 Pro Version 10.0.15063 Build 15063

Mailspring: 1.0.7-664204d8

**Comment by bengotow** [comment]
Hey! Thanks for reporting this—that's a pretty wild error code.

If you school is disabling OAuth'd apps you might want to try connecting Mailspring to your Gmail account using the IMAP/SMTP flow and an "Application Password". It's an older method for connecting accounts and maybe they're not blocking that!
 (https://support.google.com/accounts/answer/185833?hl=en).

I'll see if I can reproduce that error code.

**Comment by bengotow** [comment]
It looks like 3221225477 is `0xC0000005` in hex, which corresponds to `ACCESS_VIOLATION` on Windows. Unfortunately, that error could be just about anything. I'll see if I can narrow this down a bit—if there are any exception traces in the `mailsync-*.log` files, that could be really helpful.

**Comment by zopsi** [comment]
`69536 [2017-10-25 14:20:22.826] [background] [info] syncFolderUIDRange - ([Gmail]/All Mail, UIDs: 1 - 0, Heavy: false)
69536 [2017-10-25 14:20:47.396] [background] [error] syncFolderUIDRange - Fetching full headers for 1
69536 [2017-10-25 14:20:47.433] [background] [critical]  *** Stack trace:
 ***   pthread_timechange_handler_np
 ***   ?release@Object@mailcore@@UAEXXZ
 ***`

Later in the log 

`69580 [2017-10-25 15:09:09.282] [foreground] [info] syncFolderChangesViaCondstore - [Gmail]/All Mail: highestmodseq changed, requesting changes...
69580 [2017-10-25 15:09:09.948] [foreground] [error] Unable to fetch body for message "FedEx Shipment XXXXXXXXXXXX: Delivery scheduled for tomorrow" ([Gmail]/All Mail UID XXXXX). Error ErrorFetch`

Also a ton of these:

`21344 [2017-10-26 12:00:03.915] [background] [info] syncFolderUIDRange - ([Gmail]/All Mail, UIDs: 1 - 0, Heavy: false)
21344 [2017-10-26 12:00:16.098] [metadata] [info] Metadata delta stream closed.
21344 [2017-10-26 12:00:16.124] [metadata] [info] Encountered error: {"debuginfo":"/delta/streaming","key":"Transferred a partial file","retryable":true,"what":"Unknown exception"}. . Will retry in 300 sec.
21344 [2017-10-26 12:00:26.790] [background] [error] syncFolderUIDRange - Fetching full headers for 1`

I can give some more information when I get home or I can send you some full logs.

**Comment by zopsi** [comment]
Strange. I cleared the logs. Restarted MailSpring and no errors on my GSuite, Gmail, Fastmail.

**Comment by Zangoran** [comment]
@bengotow I looked through my logs and was unable to find anything other than just standard [info]. I also retried setting my account up using the gmail OAuth and it worked this time. This has happened before once, and when it did, I deleted Mailspring and reinstalled it and everything worked again. So I am pretty sure this will happen again. And when it does I will give you the logs. Thanks for your help!

**Comment by Zangoran** [comment]
@bengotow [https://pastebin.com/pqvUNiCW](https://pastebin.com/pqvUNiCW)

Here is the mailsync log, I removed a bunch of redundant statements as I could only upload a 512 Kb file.

**Comment by Richyread** [comment]
Just wanted to add my feedback - same error code & message, though slightly different experience;

- Updated Gmail account password
- Triggered Mailspring 're-Authenticate' option
- Background 'Gmail' screen pop-up advising reconnect
- New browser tab opens and asks to confirm authentication
- Authenticated, then asked to confirm Mailspring >> Google permissions
- Accept permissions and browser reloads Github success window
- New email from Gmail confirms Mailspring permission changes/setup

BUT
- Background 'Gmail' screen shows "sorry, we had trouble logging you in" followed by the above error code.  
- Mailspring still running but no sync activity/new emails etc.

I've quickly skimmed the log files and this is the last entry, which repeats each time I open and close Mailspring. 


![image](https://user-images.githubusercontent.com/7470850/33278605-f528abae-d393-11e7-82ae-a7480cfc0f59.png)

Richy.

**Comment by Zangoran** [comment]
@Richyread I have the same problem. It wont accept a re-authentication.

**Comment by andreyuv** [comment]
This just happened on my Windows 10 home computer.

- I have two accounts.
- Opened the computer today, the first account wouldn't sync.
- Tried reconnecting it and got that error.
- Restarted the computer. Now the first account works but the second doesn't
- Repeated the procedure with the second account (reconnected, failed, restarted computer)
- Now everything works again.

Hope this will help in some way.

**Comment by emrjackson** [comment]
I got this error after I instituted 2 step verification on my Gmail account. 

I could not re-verify even after removing the account, restarting the computer, etc. I even tried to connect as an IMAP service; no love. The only thing that worked was to manually remove the Mailspring folder from appdata>roaming. Adding my Gmail account when Mailspring first opened worked like a charm.

**Comment by bengotow** [comment]
Hey folks - I think this is relevant in @Zangoran's log file above:

```
15164 [2017-11-06 08:04:50.735] [background] [critical]  *** Stack trace:
 ***   raise
 ***   abort
 ***   terminate
 ***   wcsstr
 ***   wcsstr
 ***   RtlConvertUlongToLargeInteger
 ***   RtlConvertUlongToLargeInteger
 ***   KiUserExceptionDispatcher
 ***   _CxxThrowException
 ***   _o__CIpow
 ***   RtlSubscribeWnfStateChangeNotification
 ***
 ```

In the next release, we're also switching Mailspring to do Google OAuth using their "offline mobile app" strategy, which makes token refresh simpler and should fix those "Invalid Response Code: 400" issues that also seem to be a consistent theme here. Hopefully that will take care of this.

**Comment by TheSilvus** [comment]
I'm getting this error as well, although on a custom server. The log I'm able to access simply contains

    Waiting for Account JSON:

    Waiting for Identity JSON:

Is this a separate issue?

**Comment by bengotow** [comment]
Hey folks—I've been able to narrow this down and it'll be fixed in a small patch release tomorrow. I added better error logging for a variety of scenarios and It turns out the `3221225477` issue was due to Mailspring's sync worker being unable to open it's log file. On Windows, only one sync worker for an account can have the file open at once, and when you're re-authenticating an existing account there are two briefly.
